NIFI-7103 Adding PutAzureDataLakeStorage Processor to provide native support for Azure Data Lake Storage Gen 2

MuazmaZ opened the pull request:
#### Description of PR

Added PutAzureDataLakeStorage Processor for Azure Data lake Storage Gen 2.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [X] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

turcsanyip commented:
@MuazmaZ Thank you for the contribution. This ADLS Gen 2 processor would definitely be a valuable addition to NiFi.
However, the CI builds failed. Could you please check them in the first round?

MuazmaZ commented:
@turcsanyip added the missing dependency. All checks have passed.

sburges commented:
Functionally this looks good, I built locally and tested against an Azure Data Lake account in Azure. The success and error cases work as expected.

@MuazmaZ One small thing, it looks like something happened to the indenting in the AbstractAzureDataLakeStorageProcessor class. It would be good to clean that up for readability.

MuazmaZ commented:
Thanks @sburges for the feedback, indentation is fixed now

sburges commented:
LGTM, thanks for addressing.

MuazmaZ commented:
@pvillard31 thanks for the feedback.
- ADLS by default is encrypted with Microsoft Managed Keys. We can definitely add Key ID/Key Vault functionality for Customer Managed Keys for encryption as a feature add.
- Following Azure Blob Storage pattern we want to provide both options with Controller Service and with properties. 
I plan to add it along with permission (ACLs) as future additions. Let me know if it clarifies your questions.

pvillard31 commented:
Yes I was talking about CMEK. Totally fine with a follow-up JIRA.

Regarding the controller service, I don't see it as an option in this new processor. Here is the configuration for the Blob processor:

<img width="768" alt="Screenshot 2020-03-11 at 23 36 51" src="https://user-images.githubusercontent.com/11541012/76471064-fb251100-63f1-11ea-8096-962210554e67.png">

Where Storage Credentials is the possible reference to the controller service. CS that I could configure with:

<img width="1122" alt="Screenshot 2020-03-11 at 23 37 10" src="https://user-images.githubusercontent.com/11541012/76471099-1263fe80-63f2-11ea-985b-99cd0e8cc97a.png">

With this new processor, I don't see the Controller Service option:

<img width="783" alt="Screenshot 2020-03-11 at 23 36 41" src="https://user-images.githubusercontent.com/11541012/76470968-bf8a4700-63f1-11ea-9f91-0d80a36033a4.png">

Am I missing something?

pvillard31 commented:
Sorry I just read again your previous answer... you want to add it as a future addition. Please disregard my comments then ;)

jfrazee commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java:
```diff
@@ -0,0 +1,116 @@
+                        ACCOUNT_KEY.getDisplayName(), PROP_SAS_TOKEN.getDisplayName()));
+                }
+        final DataLakeFileSystemClient dataLakeFileSystemClient = storageClient.getFileSystemClient(fileSystem);
+        final DataLakeDirectoryClient directoryClient = dataLakeFileSystemClient.getDirectoryClient(directory);
```
Are there any exceptions that can be thrown out of this part of the code and should we be handling the `IllegalArgumentException`? I don't think whether `ACCOUNT_KEY` or `PROP_SAS_TOKEN` are defined will change between retries. It might be worth adding an IT test for that scenario to sort out whether it matters.

MuazmaZ replied:
Thanks @jfrazee , added the IllegalArgumentException to capture the scenario.

pvillard31 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml:
```diff
@@ -74,7 +74,7 @@
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-core</artifactId>
-            <version>${jackson.version}</version>
+            <version>2.10.1</version>
```
Is this change required? If yes, can we use the latest (2.10.3)?
Ideally, I would try to remain consistent but I agree that changing the Jackson version in the root pom can be quite impacting...

MuazmaZ replied:
Yes, it's required to use the latest sdk release. 2.10.1 matches the azure core sdk. I would stick to the azure version as compared to the latest. Let me know if it makes sense.

MuazmaZ replied:
@pvillard31 could you let me know if this PR is good to be merged or you have additional comments?

pvillard31 replied:
No, it's good on my side. If @turcsanyip is a +1, happy to merge.

MuazmaZ commented:
@turcsanyip are you good with this? I would like to merge soon.

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java:
```diff
@@ -0,0 +1,121 @@
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+            session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl().toString(), transferMillis);
+        } catch (IllegalArgumentException e) {
+            getLogger().error("Failed to create {}, due to {}", new Object[]{flowFile, e}, e);
```
"Failed to upload {FlowFile}" message may be more adequate. The FF already exists, so "Failed to create {FlowFile}" sounds a bit misleading for me.

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java:
```diff
@@ -0,0 +1,154 @@
+            AbstractAzureDataLakeStorageProcessor.PROP_SAS_TOKEN, AbstractAzureDataLakeStorageProcessor.FILE));
+
+    private static final Set <Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet <>(
+        asList(
```
Could you please use either `Arrays.asList` or `asList` (with static import) both in lines 113 and 118?

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java:
```diff
@@ -0,0 +1,121 @@
+            session.transfer(flowFile, REL_SUCCESS);
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+            session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl().toString(), transferMillis);
+        } catch (IllegalArgumentException e) {
```
All exceptions should be caught here. Eg. DataLakeStorageException coming from directoryClient.createFile(fileName) in line 93. Also exceptions from fileClient.append(in, 0, length) could be handled here.
So I would modify the class to Exception.

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java:
```diff
@@ -0,0 +1,154 @@
+
+    public static final PropertyDescriptor FILESYSTEM = new PropertyDescriptor.Builder().name("Filesystem-name")
+        .displayName("Filesystem Name")
+        .description("Name of the Azure Storage File System, It is assumed to be already existing")
```
Typo: `','` at the end of the first sentence. No `'.'` at the end of the second sentence.
The same for the Directory property.

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java:
```diff
@@ -0,0 +1,154 @@
+
+    public static final PropertyDescriptor DIRECTORY = new PropertyDescriptor.Builder().name("Directory-name")
+        .displayName("Directory Name")
+        .description("Name of the Azure Storage Directory, It is assumed to be already existing")
```
I think the non existing directories will be created. At least I did not get any errors.

markap14 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java:
```diff
@@ -0,0 +1,121 @@
+            final Map<String, String> attributes = new HashMap<>();
+            final DataLakeFileClient fileClient = directoryClient.createFile(fileName);
+            if (length > 0) {
+                try (final InputStream rawIn = session.read(flowFile); final BufferedInputStream in = new BufferedInputStream(rawIn)) {
```
`InputStreamCallback` has existed from the beginning. The `InputStream ProcessSession.read(FlowFIle)` is a newer mechanism. Neither is necessarily preferred over the other. But the newer method, as used here, is generally easier to read, so I'd lean toward preferring that.

turcsanyip replied:
@markap14 Thanks for the clarification. Then the current solution is appropriate, no need to change it.

MuazmaZ commented:
@turcsanyip Thanks for your feedback. I have addressed all the comments. For high-level feedback:
- getRootDirectoryClient() function is not public in DataLakeFileSystemClient, I have reached out to the SDK team on this.
- Overwriting a file is also a follow-up with the SDK team as API supports a conditional If not match (https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create)
-  ACLs will be a follow-up task, already on the list. Some known issues need to be verified.

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java:
```diff
@@ -0,0 +1,96 @@
+            session.transfer(flowFile, REL_SUCCESS);
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+            session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl(), transferMillis);
+        } catch (IOException | StorageErrorException | IllegalArgumentException e) {
```
I'd recommend to catch `Exception` here.
I managed to throw `DataLakeStorageException` from the processor (eg. via specifying a non-existing filesystem or an invalid directory name on the UI). In this case the flowfile does not go to failure, but always goes back to the input queue of the processor which is not the expected behaviour I think (a wrong flowfile could hang up the processor, the other flowfiles would not be processed).
`DataLakeStorageException` is just an example I think so I would not add it as a fourth exception to be caught but rather catch `Exception`.
Or catch `(IOException | AzureException | IllegalArgumentException)`, but I'm afraid there can be other non-expected `RuntimeException-s` coming out from the library, so `Exception` seems to me the safest solution.

MuazmaZ commented:
@turcsanyip confirmed with the SDK team that getDirectoryClient(“”) is valid. Made the change in property to allow empty string.
Removed PROP_ for SAS_TOKEN
Updated to catch Exception and tested the scenario. I was able to see the failure files with a non-existing filesystem.

ravitejatvs commented:
@MuazmaZ I'm trying to push data to ADLS Gen2. But everytime, I get an exception that says `Failed to create file on Azure Data Lake Storage: com.azure.storage.file.datalake.models.DataLakeStorageException: Status code 400, "{"error":{"code":"InvalidUri","message":"The request URI is invalid.\nRequestId:3d7bfb40-b01f-0050-19d3-81c973000000\nTime:2020-09-03T09:17:20.1611855Z"}}"`. 

I am sure that the details provided are correct but still I am getting this. I also tried sending data to Storage Queue using the corresponding processor and that works fine. 

Please let me know if you need more details.
